pkg/operator/controller/status.go

```diff
@@ -132,100 +218,79 @@ func computeOperatorDegradedCondition(conditions []configv1.ClusterOperatorStatu
-	ingresses []operatorv1.IngressController) []configv1.ClusterOperatorStatusCondition {
+func computeOperatorAvailableCondition(oldCondition *configv1.ClusterOperatorStatusCondition,
+	allIngressesAvailable bool) configv1.ClusterOperatorStatusCondition {
 	availableCondition := &configv1.ClusterOperatorStatusCondition{
-		Type:   configv1.OperatorAvailable,
-		Status: configv1.ConditionUnknown,
+		Type: configv1.OperatorAvailable,
 	}
-	numIngresses := len(ingresses)
-	var ingressesAvailable int
-	for _, ing := range ingresses {
-		for _, c := range ing.Status.Conditions {
-			if c.Type == operatorv1.IngressControllerAvailableConditionType && c.Status == operatorv1.ConditionTrue {
-				ingressesAvailable++
-				break
-			}
-		}
-	}
-	if numIngresses == ingressesAvailable {
+
+	if allIngressesAvailable {
 		availableCondition.Status = configv1.ConditionTrue
 		availableCondition.Message = ingressesEqualConditionMessage
 	} else {
 		availableCondition.Status = configv1.ConditionFalse
 		availableCondition.Reason = "IngressUnavailable"
-		availableCondition.Message = fmt.Sprintf(
-			"%d ingress controllers available, want %d",
-			ingressesAvailable, numIngresses)
+		availableCondition.Message = fmt.Sprintf("all ingress controllers are not available")
 	}
 
-	return setOperatorStatusCondition(conditions, availableCondition)
+	return setLastTransitionTime(availableCondition, oldCondition)
 }
 
-// setOperatorStatusCondition returns a slice of Operator status conditions as
-// a result of setting the specified condition in the given slice of conditions.
-func setOperatorStatusCondition(oldConditions []configv1.ClusterOperatorStatusCondition, condition *configv1.ClusterOperatorStatusCondition) []configv1.ClusterOperatorStatusCondition {
-	condition.LastTransitionTime = metav1.Now()
-
-	newConditions := []configv1.ClusterOperatorStatusCondition{}
-
-	found := false
-	for _, c := range oldConditions {
-		if condition.Type == c.Type {
-			if condition.Status == c.Status &&
-				condition.Reason == c.Reason &&
-				condition.Message == c.Message {
-				return oldConditions
-			}
-
-			found = true
-			newConditions = append(newConditions, *condition)
-		} else {
-			newConditions = append(newConditions, c)
-		}
-	}
-	if !found {
-		newConditions = append(newConditions, *condition)
+// setLastTransitionTime sets LastTransitionTime for the given condition.
+// If the condition has changed, it will assign a new timestamp otherwise keeps the old timestamp.
+func setLastTransitionTime(condition, oldCondition *configv1.ClusterOperatorStatusCondition) configv1.ClusterOperatorStatusCondition {
+	if oldCondition != nil && condition.Status == oldCondition.Status &&
+		condition.Reason == oldCondition.Reason && condition.Message == oldCondition.Message {
+		condition.LastTransitionTime = oldCondition.LastTransitionTime
+	} else {
+		condition.LastTransitionTime = metav1.Now()
 	}
 
-	return newConditions
+	return *condition
```

Comment:
We can get this in follow-up if you prefer, but it is strange that `setLastTransitionTime` mutates `condition` and then returns a value (i.e., a copy).  It would make more sense for `setLastTransitionTime` to either copy and then mutate that or not to return a copy at all.